pkg/keycloak/phaseHandler.go

```diff
@@ -100,15 +100,15 @@ func (ph *phaseHandler) Accepted(sso *v1alpha1.Keycloak) (*v1alpha1.Keycloak, er
 // Note we rely on imagestreams that the operator itself does not install (these are installed currently by the installer in the openshift ns )
 func (ph *phaseHandler) Upgrade(sso *v1alpha1.Keycloak) (*v1alpha1.Keycloak, error) {
 	cpSSO := sso.DeepCopy()
-	dc, err := ph.ocDCClient.DeploymentConfigs(cpSSO.Namespace).Get(SSO_APPLICATION_NAME, v12.GetOptions{})
-	if err != nil {
-		return cpSSO, errors.Wrap(err, "failed to get current deploymentconfig for sso")
-	}
 	if !CanUpgrade(cpSSO.Status.Version) || (cpSSO.Status.Phase != v1alpha1.PhaseUpgrading && cpSSO.Status.Phase != v1alpha1.PhaseReconcile) {
```

Comment:
I believe the logic in the `CanUpgrade` function may need to be modified. I may be missing some context, it looks like it will return try for an upgrade if the version in the CR matches v7.2.x.GA. So if the version was at v7.2.2.GA for example and stays the same, in the next reconcile loop, it would still return true and trigger the upgrade path again. Or if for some reason the version got downgraded to 7.2.1.GA, it would return true, although this may be intended depending on the upgrade logic.

Reply:
Yes it then marks the CR as in the phase upgrading and the logic proceeds from there

Reply:
So the logic is 
if we cant upgrade or an upgrade is already in progress or the initial installation hasn't finished ignore and return

If we get past that check and it is in the reconcile phase (IE already installed) then mark the phase as upgrading

Then next time through do the actual upgrade.

Reply:
Which makes sense. What Phase does the status get set to after the initial upgrade or installation?

This Upgrade function gets called on each reconcile loop from https://github.com/integr8ly/keycloak-operator/blob/1259cb0d658310fa5a63b2574b0df2b26e66c182/pkg/keycloak/keycloak.go#L125. The `CanUpgrade` will return true anytime the version in the spec matches `v7.2.x.GA`.
https://github.com/integr8ly/keycloak-operator/blob/6963509cbbfec415f5fcf9a4bf3e7b644e5322a1/pkg/keycloak/upgrade.go#L10-L14

The rest of the logic in this function will get executed even if the phase is not in `PhaseReconcile`. Do you think this may cause any issues? Is it possible for the phase to return to `PhaseReconcile` at any point, maybe due to an edit of the CR outside of the version field?

Reply:
I can't imaging changing the phase or anything in the status block would not be a supported action on any k8s resource. 
However if someone set it to phase reconcile it would come back through the function and be set back to upgrading. Then it would come back through again and attempt the upgrade again.
If it got partially through the upgrade it has defensive checks to check is the deployment upgraded and is the service upgraded.
The last thing it does is update the version field and sets the the phase back to reconcile

Reply:
It is good there are defensive checks in place. It certainly sounds like there is no issue with the way it is working currently because of this. Do you think it is worth modifying the logic in `CanUpgrade` to check the version in the CR is a higher version than the version currently deployed? This would rule out any potential issues in the future if additional logic was added to the `Upgrade` function without a defensive check.

Reply:
No I don't really see the benefit here. The operator controls when that status block and so decides when that version is specified. Using that to indicate the version of the component that is deployed seems reasonable. At least it does for this release. Perhaps when we do the next upgrade it could be revisited. Particularly as what we are trying to do here is solve the installation bug.

Reply:
Absolutely, this discussion is definitely not related to this PR. Thank you very much for your responses. Approving now.